### DISCUSSION TITLE
How to Conditionally set/change FastAPI response model for route

### DISCUSSION THREAD

**Discussion by Dijkie85** [fastapi/fastapi#6765] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

I'm trying to return a list of objects of type Company, including only "approved" ones, and with more or less attributes depending on whether the user requesting the list is a superuser or a regular user. This is my code so far:

```Python
# Route
@router.get("/", response_model=List[schema.CompanyRegularUsers])
def get_companies(db: Session = Depends(get_db), is_superuser: bool = Depends(check_is_superuser)):
    """
    If SU, also include sensitive data.
    """
    if is_superuser:
        return crud.get_companies_admin(db=db)
    return crud.get_companies_user(db=db)

# CRUD functions (not the DRYest code, I know...):

def get_companies_admin(db: Session, skip: int = 0, limit: int = 100):
    db_companies = db.query(model.Company).offset(skip).limit(limit).all()
    return db_companies

def get_companies_user(db: Session, skip: int = 0, limit: int = 100):
    db_companies = db.query(model.Company).filter(model.Company.is_approved == True).all()
    return db_companies

```

### Description

The path operation correctly returns the objects according to request (ie., only `is_approved=True` companies if a regular request, and both `is_approved=True` and `is_approved=False` if requested by a superuser. Problem is, both cases use `schema.CompanyRegularUsers` (as set by `response_model`) and I'd like to use `schema.CompanySuperusers` when SU's make the request.

How can I achieve that feature? I.e, is there a way to conditionally set the response_model property of the decorator function?

I've tried using JSONResponse with Pydantic's `schema.CompanySuperusers.from_orm()`, but it won't work with a list of Companies. Also tried Pydantic's `parse_obj_as`, but it seems to only work from JSON to model, and not the reverse, which is what I'm looking for..

**Comment by Dijkie85** [comment, marked as answer]
Solved as explained here, in case anyone finds the answer useful:


https://stackoverflow.com/a/66948376/12104222

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
